nav2_collision_monitor/src/collision_detector_node.cpp

```diff
@@ -0,0 +1,317 @@
+{
+  RCLCPP_INFO(get_logger(), "Activating");
+
+  // Activating lifecycle publisher
+  state_pub_->on_activate();
+
+  // Activating polygons
+  for (std::shared_ptr<Polygon> polygon : polygons_) {
+    polygon->activate();
+  }
+
+  // Creating timer
+  timer_ = this->create_timer(
+      std::chrono::duration<double>{1.0 / frequency_},
+      std::bind(&CollisionDetector::process, this));
+
+  // Creating bond connection
+  createBond();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+CollisionDetector::on_deactivate(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Deactivating");
+
+  // Deactivating lifecycle publishers
+  state_pub_->on_deactivate();
+
+  // Deactivating polygons
+  for (std::shared_ptr<Polygon> polygon : polygons_) {
+    polygon->deactivate();
+  }
+
+  // Resetting timer
+  timer_.reset();
+
+  // Destroying bond connection
+  destroyBond();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+CollisionDetector::on_cleanup(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Cleaning up");
+
+  state_pub_.reset();
+
+  polygons_.clear();
+  sources_.clear();
+
+  tf_listener_.reset();
+  tf_buffer_.reset();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+CollisionDetector::on_shutdown(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Shutting down");
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+bool CollisionDetector::getParameters()
+{
+  std::string base_frame_id, odom_frame_id;
+  tf2::Duration transform_tolerance;
+  rclcpp::Duration source_timeout(2.0, 0.0);
+
+  auto node = shared_from_this();
+
+  nav2_util::declare_parameter_if_not_declared(
+    node, "frequency", rclcpp::ParameterValue(10.0));
+  frequency_ = get_parameter("frequency").as_double();
+  nav2_util::declare_parameter_if_not_declared(
+    node, "base_frame_id", rclcpp::ParameterValue("base_footprint"));
+  base_frame_id = get_parameter("base_frame_id").as_string();
+  nav2_util::declare_parameter_if_not_declared(
+    node, "odom_frame_id", rclcpp::ParameterValue("odom"));
+  odom_frame_id = get_parameter("odom_frame_id").as_string();
+  nav2_util::declare_parameter_if_not_declared(
+    node, "transform_tolerance", rclcpp::ParameterValue(0.1));
+  transform_tolerance =
+    tf2::durationFromSec(get_parameter("transform_tolerance").as_double());
+  nav2_util::declare_parameter_if_not_declared(
+    node, "source_timeout", rclcpp::ParameterValue(2.0));
+  source_timeout =
+    rclcpp::Duration::from_seconds(get_parameter("source_timeout").as_double());
+
+  if (!configurePolygons(base_frame_id, transform_tolerance)) {
+    return false;
+  }
+
+  if (!configureSources(base_frame_id, odom_frame_id, transform_tolerance, source_timeout)) {
+    return false;
+  }
+
+  return true;
+}
+
+bool CollisionDetector::configurePolygons(
```

Comment:
I propose in this function to set the action_type of the polygons to DO_NOTHING and not include the action_type yaml parameter in the docs examples. I think it would be confusing to offer an action_type yaml option when there is really only one choice.

The downside is that this makes it harder to put configurePolygons in a common utils.cpp file.

What do you think?

Reply:
I don't think there's a problem to set it here to that value. But I would recommend logging if its NOT either empty or `notify` (or whatever rename it) so users know there's an issue with what they're requesting the system to accomplish.

For the sake of reusing code in general utilities, it might not make sense to remove that logic here but keep the same as much as possible to be shared, then application specific code after the shared function for validation.